DEVELOPMENT.md

```diff
@@ -0,0 +1,223 @@
+* Use `get...` method names for pure methods that only return the value without doing any heavy workload like accessing git repository. <br/>
+  Use `derive...` method names for methods that actually compute their result and/or can return a different value every time when accessed.
+* Non-obvious method params that have values like `false`, `true`, `0`, `1`, `null`, `""` should be preceded with a `/* comment */ `
+  containing the name of the param.
+
+
+## Rebuild the CI base image
+
+To push the rebuilt image, you need a write access to [`gitmachete` organization on Docker Hub](https://hub.docker.com/orgs/gitmachete).
+
+```
+docker build -t gitmachete/intellij-plugin-ci .
+docker push gitmachete/intellij-plugin-ci
+```
+
+
+## Versioning
+
+We follow [Semantic versioning](https://semver.org/) for the plugin releases:
+
+* MAJOR version must be bumped for each plugin release that stops supporting any IDEA build (typically when `sinceBuild` is increased). <br/>
+  This does not apply to 0->1 major version transition, which is going to happen when the plugin's compatibility range is considered stable.
+* MINOR version must be bumped for each plugin release that either adds a new user-facing feature
+  or starts supporting a new quarterly (`year.number`) IDEA build (typically when `untilBuild` is increased).
+* PATCH version must be bumped for each plugin release that adds no new user-facing features
+  and doesn't change the range of supported IDEA builds.
+
+### Sample sequence of versions between releases
+
+After a release e.g. `1.0.3`, subsequent PRs merged to `develop` might change `PROSPECTIVE_RELEASE_VERSION`
+in [version.gradle](version.gradle) in the following way:
+1. `1.0.4` (bugfix PR)  - the first PR merged to develop after the release must bump `PROSPECTIVE_RELEASE_VERSION` since of course the prospective release won't be `1.0.3` anymore
+1. `1.0.4` (bugfix PR)  - even if a new set of patch-level changes has been added on the PR, the released version is still going to be `1.0.4` (not `1.0.5`)
+1. `1.1.0` (feature PR) - since we've just added a new feature, the new release won't be a PATCH-level anymore, but MINOR-level one
+1. `1.1.0` (bugfix PR)  - even if a new feature has been added on the PR, the released version is still going to be `1.1.0` (not `1.2.0`)
+1. `2.0.0` (breaking change PR)
+1. `2.0.0` (feature PR) - again, still `2.0.0` and not e.g. `2.1.0`
+1. `2.0.0` (bugfix PR)
+1. `2.0.0` (release PR) - finally releasing as major release; as a consequence, `1.0.4` and `1.1.0` never actually gets released
+
+### IDE supported versions
+
+Since we cannot skip `untilBuild` field in plugin build configuration (see related [issue](https://github.com/VirtusLab/git-machete-intellij-plugin/issues/460) and [YouTrack ticket](https://youtrack.jetbrains.com/issue/IJSDK-888)) the most reasonable approach is to bump `untilBuild` to `X.*` when new `X` EAP or RC version will be released.
```

Comment:
Nit:
```suggestion
Since we cannot skip `untilBuild` field in plugin build configuration
(see related [issue](https://github.com/VirtusLab/git-machete-intellij-plugin/issues/460) 
and [YouTrack ticket](https://youtrack.jetbrains.com/issue/IJSDK-888)),
the most reasonable approach is to bump `untilBuild` to `X.*` when the new `X` EAP or RC version is released.
```

Reply:
Rule of the thumb: there's never future tense after `when`, and almost never after `if`